### PR TITLE
Introduce descriptive types to refer to task input and output files and directories

### PR DESCRIPTION
**Pull request by adammurdoch**
A solution for #2290 and #2305, introducing various APIs to define and reference lazily calculated directory and file locations, and use these to define and wire together task input and output files and directories.

### CODE REVIEW

**Comment by oehme** [comment]
This looks like a good step forward. 

- It neatly solves the performance problem of eagerly canonicalizing files in code like this:

```
test.report.html.destinationDir = file("$reportsDir/someName")
```

Which could now be something like (Ignoring backwards compatibility concerns):

```
test.report.html.destinationDir = layout.reportsDir.dir("someName")
```

- Fitting this onto our existing types might become awkward, since we'd have to find new names for all the getters. 

- I also wonder whether we should have some DSL decoration that allows me to assign a `Provider<File>` to a "legacy" task that wants a `File`.

- We should explain this concept in the user guide/plugin developer guide.

File: subprojects/core/src/main/java/org/gradle/api/file/DirectoryVar.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Represents some configurable directory location, whose value is mutable and is not necessarily currently known until later.
+ * <p>
+ * <b>Note:</b> This interface is not intended for implementation by build script or plugin authors. An instance of this class can be created
+ * using the {@link ProjectLayout#newDirectoryVar()} method.
+ *
+ * @since 4.1
+ */
+@Incubating
+public interface DirectoryVar extends Provider<Directory>, PropertyState<Directory> {
```
Comment (lacasseio) [inline subprojects/core/src/main/java/org/gradle/api/file/DirectoryVar.java]: I'm a bit uneasy if the name for this interface (and `RegularFileVar`). I fail to see the reason of using suffix `Var` to the interface name. I feel like `ConfigurableDirectory` (and `ConfigurableRegularFile`) would be more appropriate. I base myself on existing name such as `ConfigurableFileCollection` and `ConfigurableFileTree`.

Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/api/file/DirectoryVar.java]: Thanks for reviewing. I don't love these names either, but I don't love the `ConfigurableThing` names either.

**Comment by adammurdoch** [comment]
@oehme Thanks for the feedback.

> Fitting this onto our existing types might become awkward, since we'd have to find new names for all the getters.

I think this is something that we don't have a good answer for yet. It's already been tricky to come up with new names for the properties on the native tasks.

> I also wonder whether we should have some DSL decoration

Yes, I plan to add something there.

> We should explain this concept in the user guide/plugin developer guide.

Absolutely.

**Comment by alkemist** [comment]
@adammurdoch I'm only seeing one commit on this PR now and it doesn't seem to encompass the topic. Did you force push the wrong thing by chance?

**Comment by adammurdoch** [comment]
@alkemist I did indeed, sorry. The changes have already been merged into master and are here: https://github.com/gradle/gradle/compare/93559d6...261adcb

**Pull request closed without merging**